src/compiler/checker.ts

```diff
@@ -5434,14 +5435,14 @@ namespace ts {
         }
 
         function createSignature(declaration: SignatureDeclaration, typeParameters: TypeParameter[], thisParameter: Symbol | undefined, parameters: Symbol[],
-            resolvedReturnType: Type, typePredicate: TypePredicate, minArgumentCount: number, hasRestParameter: boolean, hasLiteralTypes: boolean): Signature {
+            resolvedReturnType: Type | undefined, resolvedTypePredicate: TypePredicate | undefined | "pending", minArgumentCount: number, hasRestParameter: boolean, hasLiteralTypes: boolean): Signature {

```

Comment:
Please be aware that if we call `createSignature` once with a `TypePredicate` and then call `createSignature` again later with `"pending"` that it will result in a V8 compiling a second copy of `createSignature` due to the parameter being polymorphic (string vs. object).

We also end up making `Signature` more polymorphic. This could result in degraded performance so I would recommend you run benchmarks before and after this change to ensure this does not cause a regression.

As an alternative, I would suggest that you define something like a `const unresolvedTypePredicate = <IdentifierTypePredicate>{ kind: TypePredicateKind.Identifier, parameterName: "pending", parameterIndex: 0 };` to act as your `"pending"` sentinel value as it will reduce polymorphism.